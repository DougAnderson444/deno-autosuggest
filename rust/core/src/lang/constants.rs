use super::CharClass;
use CharClass::{Consonant, Vowel};

pub const CHAR_CLASSES_LATIN: &[(CharClass, char)] = &[
    (Consonant, 'b'),
    (Consonant, 'c'),
    (Consonant, 'd'),
    (Consonant, 'f'),
    (Consonant, 'g'),
    (Consonant, 'h'),
    (Consonant, 'j'),
    (Consonant, 'k'),
    (Consonant, 'l'),
    (Consonant, 'm'),
    (Consonant, 'n'),
    (Consonant, 'p'),
    (Consonant, 'q'),
    (Consonant, 'r'),
    (Consonant, 's'),
    (Consonant, 't'),
    (Consonant, 'v'),
    (Consonant, 'w'),
    (Consonant, 'x'),
    (Consonant, 'z'),
    (Vowel, 'a'),
    (Vowel, 'e'),
    (Vowel, 'i'),
    (Vowel, 'o'),
    (Vowel, 'u'),
    (Vowel, 'y'),
];
