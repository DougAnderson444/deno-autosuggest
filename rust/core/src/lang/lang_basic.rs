#![allow(dead_code)]

use super::Lang;
use super::constants::CHAR_CLASSES_LATIN;

pub fn lang_basic() -> Lang {
    let mut lang = Lang::new();
    for &(class, ch) in CHAR_CLASSES_LATIN { lang.add_char_class(ch, class); }
    lang
}
