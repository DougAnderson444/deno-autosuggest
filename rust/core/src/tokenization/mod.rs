mod word;
mod text;

use crate::lang::{Lang, Chars, CharPattern};
pub use word::Word;
pub use text::Text;


pub fn tokenize_query(source: &str, lang: &Option<Lang>) -> Text<Vec<char>> {
    Text::from_str(source)
        .normalize(lang)
        .fin(false)
        .split(&[Chars::Whitespaces, Chars::Control, Chars::Punctuation])
        .strip(&[Chars::NotAlphaNum])
        .lower()
        .mark_pos(lang)
        .stem(lang)
}


pub fn tokenize_record(source: &str, lang: &Option<Lang>) -> Text<Vec<char>> {
     Text::from_str(source)
        .normalize(lang)
        .split(&[Chars::Whitespaces, Chars::Control, Chars::Punctuation])
        .strip(&[Chars::NotAlphaNum])
        .lower()
        .mark_pos(lang)
        .stem(lang)
}
