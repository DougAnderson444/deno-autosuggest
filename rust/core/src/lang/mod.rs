mod normalize;
mod lang_english;
mod lang_german;
mod lang_portuguese;
mod lang_russian;
mod lang_spanish;

pub use lang_german::lang_german;
pub use lang_english::lang_english;
pub use lang_spanish::lang_spanish;
pub use lang_portuguese::lang_portuguese;
pub use lang_russian::lang_russian;

use std::cell::RefCell;
use fnv::{FnvHashMap as HashMap};
use rust_stemmers::Stemmer;
use crate::utils::to_vec;
use crate::tokenization::PartOfSpeech;
use normalize::Normalize;


const BUFFER_CAPACITY: usize = 20;


pub struct Lang {
    stemmer:      Option<Stemmer>,
    pos_map:      HashMap<Vec<char>, PartOfSpeech>,
    compose_map:  HashMap<Vec<char>, Vec<char>>,
    reduce_map:   HashMap<Vec<char>, Vec<char>>,
    stem_buffer:  RefCell<String>,
    norm_buffer1: RefCell<Vec<char>>,
    norm_buffer2: RefCell<Vec<char>>,
}


impl Lang {
    pub fn new() -> Self {
        Self {
            stemmer:      None,
            pos_map:      HashMap::default(),
            compose_map:  HashMap::default(),
            reduce_map:   HashMap::default(),
            stem_buffer:  RefCell::new(String::with_capacity(BUFFER_CAPACITY)),
            norm_buffer1: RefCell::new(Vec::with_capacity(BUFFER_CAPACITY)),
            norm_buffer2: RefCell::new(Vec::with_capacity(BUFFER_CAPACITY)),
        }
    }

    pub fn set_stemmer(&mut self, stemmer: Option<Stemmer>) {
        self.stemmer = stemmer;
    }

    pub fn add_pos(&mut self, word: &str, pos: PartOfSpeech) {
        let source   = to_vec(word);
        let composed = self.unicode_compose(&source).unwrap_or(source);
        let reduced  = self.unicode_reduce(&composed);

        self.pos_map.insert(composed, pos);

        if let Some((_, reduced)) = reduced {
            self.pos_map.insert(reduced, pos);
        }
    }

    pub fn add_unicode_composition(&mut self, from: &str, to: &str) {
        self.compose_map.insert(to_vec(from), to_vec(to));
    }

    pub fn add_unicode_reduction(&mut self, from: &str, to: &str) {
        self.reduce_map.insert(to_vec(from), to_vec(to));
    }

    pub fn stem(&self, word: &[char]) -> usize {
        if let Some(stemmer) = &self.stemmer {
            let buffer = &mut *self.stem_buffer.borrow_mut();
            buffer.clear();
            buffer.extend(word.iter());
            let stem = stemmer.stem(&buffer);
            stem.chars().count()
        } else {
            word.len()
        }
    }

    pub fn get_pos(&self, word: &[char]) -> Option<PartOfSpeech> {
        self.pos_map.get(word).cloned()
    }

    pub fn unicode_compose(&self, word: &[char]) -> Option<Vec<char>> {
        let buffer = &mut *self.norm_buffer1.borrow_mut();
        buffer.clear();

        for (_, norm_chunk) in Normalize::new(word, &self.compose_map) {
            buffer.extend(norm_chunk);
        }

        if &buffer[..] == word {
            None
        } else {
            Some(buffer.clone())
        }
    }

    pub fn unicode_reduce(&self, word: &[char]) -> Option<(Vec<char>, Vec<char>)> {
        let buffer1 = &mut *self.norm_buffer1.borrow_mut();
        let buffer2 = &mut *self.norm_buffer2.borrow_mut();
        buffer1.clear();
        buffer2.clear();

        for (word_chunk, norm_chunk) in Normalize::new(word, &self.reduce_map) {
            buffer1.extend(word_chunk);
            buffer2.extend(norm_chunk);
            for _ in 0 .. norm_chunk.len() - word_chunk.len() {
                buffer1.push('\0');
            }
        }

        if &buffer2[..] == word {
            None
        } else {
            Some((buffer1.clone(), buffer2.clone()))
        }
    }
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::utils::to_vec;
    use crate::tokenization::PartOfSpeech;
    use super::Lang;

    fn get_lang() -> Lang {
        let mut lang = Lang::new();
        lang.add_unicode_composition("ó", "ó");
        lang.add_unicode_reduction("ó", "o");
        lang.add_unicode_reduction("õ", "oo");
        lang.add_pos("fóo", PartOfSpeech::Particle);
        lang
    }

    #[test]
    fn unicode_compose_nfc() {
        let input  = to_vec("foóbar");
        let output = get_lang().unicode_compose(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn unicode_compose_nfd() {
        let input  = to_vec("foóbar");
        let output = get_lang().unicode_compose(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn unicode_reduce_reduced() {
        let input  = to_vec("foobar");
        let output = get_lang().unicode_reduce(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn unicode_reduce_nfc() {
        let input  = to_vec("foóbar");
        let output = get_lang().unicode_reduce(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn unicode_reduce_fill0() {
        let input  = to_vec("fõbar");
        let output = get_lang().unicode_reduce(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn get_pos_source() {
        let input = to_vec("fóo");
        let pos   = get_lang().get_pos(&input);
        assert_eq!(pos, Some(PartOfSpeech::Particle));
    }

    #[test]
    fn get_pos_reduced() {
        let input = to_vec("foo");
        let pos   = get_lang().get_pos(&input);
        assert_eq!(pos, Some(PartOfSpeech::Particle));
    }

    #[test]
    fn get_pos_unknown() {
        let input = to_vec("bar");
        let pos   = get_lang().get_pos(&input);
        assert_eq!(pos, None);
    }
}