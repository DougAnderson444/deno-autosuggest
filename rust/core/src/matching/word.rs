use crate::tokenization::Word;
use super::WordMatch;
use super::damlev::DamerauLevenshtein;
use super::jaccard::Jaccard;


const LENGTH_THRESHOLD:  f64 = 0.26;
const JACCARD_THRESHOLD: f64 = 0.41;
const DAMLEV_THRESHOLD:  f64 = 0.26;

thread_local! {
    static DAMLEV:  DamerauLevenshtein = DamerauLevenshtein::new();
    static JACCARD: Jaccard<char>      = Jaccard::new();
}


pub fn word_match(rword: &Word, qword: &Word, rchars: &[char], qchars: &[char]) -> Option<WordMatch> {
    if qword.is_empty() || rword.is_empty() {
        return None;
    }
    if !length_check(rword, qword) {
        return None;
    }
    if !jaccard_check(rword, qword, rchars, qchars) {
        return None;
    }

    let rchars = rword.view(rchars);
    let qchars = qword.view(qchars);
    let mut best_match: Option<WordMatch> = None;

    DAMLEV.with(|damlev| {
        damlev.distance(qchars, rchars);
        let dists = &*damlev.dists.borrow();

        let left  = if qword.fin { max!(qword.stem, rword.stem) } else { qword.stem } - 1;
        let right = max!(qword.len(), rword.len()) + 1;

        if right <= left { return; }

        let range = (left .. right).rev();

        for rlen in range.clone() {
            for qlen in range.clone() {
                // Out of bounds.
                if qlen > qword.len() { continue; }
                if rlen > rword.len() { continue; }
                // Left margin is for insertion/deletion, not for both prefixes at the same time.
                if rlen == left && qlen == left  { continue; }
                // Compare full words only if query is finished.
                if qword.fin && rlen < rword.stem { break; }
                if qword.fin && qlen < qword.stem { break; }
                // Words with 2+ insertions/deletions are mismatched by default.
                if (qlen as isize - rlen as isize).abs() > 1 { continue; }

                let dist = dists.get(qlen + 1, rlen + 1);
                let rel  = dist as f64 / max!(qlen, rlen, 1) as f64;

                if rel > DAMLEV_THRESHOLD { continue; }

                best_match = best_match
                    .take()
                    .filter(|m| m.typos <= dist)
                    .or_else(|| Some(WordMatch::new(
                        qword,
                        rword,
                        qlen,
                        rlen,
                        dist,
                    )));

                if dist == 0 {
                    break;
                }
            }
        }
    });

    best_match
}


pub fn length_check(rword: &Word, qword: &Word) -> bool {
    let qlen  = qword.len();
    let rlen  = if qword.fin { rword.len() } else { min!(qlen, rword.len()) };

    if qlen <= 1 || rlen <= 1 {
        return qlen == rlen;
    }

    let long  = max!(qlen, rlen);
    let short = min!(qlen, rlen);
    let dist  = 1.0 - (short as f64 / long as f64);

    dist < LENGTH_THRESHOLD
}


pub fn jaccard_check(rword: &Word, qword: &Word, rchars: &[char], qchars: &[char]) -> bool {
    let rchars = rword.view(rchars);
    let qchars = qword.view(qchars);
    let rslice = if qword.fin { &rchars } else { &rchars[.. min!(qword.len(), rword.len())] };
    let dist   = JACCARD.with(|j| j.rel_dist(&rslice, &qchars));
    dist < JACCARD_THRESHOLD
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::utils::to_vec;
    use crate::tokenization::{Chars, Word, Text};
    use crate::lang::lang_english;
    use super::{length_check, jaccard_check, word_match};


    fn text(s: &str) -> Text<Vec<char>> {
        Text::from_str(s).split(&Chars::Whitespaces)
    }

    #[test]
    fn length_check_len_4() {
        let sample = [
            (false, "m"),
            (false, "ma"),
            (true,  "mai"),
            (true,  "mail"),
            (true,  "mailb"),
            (false, "mailbo"),
            (false, "mailbox"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mail");
            let qtext  = text(query);
            let result = length_check(&rtext.words[0], &qtext.words[0]);
            assert_eq!(result, expect, "Failed length_check(\"mail\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn length_check_len_7() {
        let sample = [
            (false, "m"),
            (false, "ma"),
            (false, "mai"),
            (false, "mail"),
            (false, "mailb"),
            (true,  "mailbo"),
            (true,  "mailbox"),
            (true,  "mailboxe"),
            (true,  "mailboxes"),
            (false, "mailboxess"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mailbox");
            let qtext  = text(query);
            let result = length_check(&rtext.words[0], &qtext.words[0]);
            assert_eq!(result, expect, "Failed length_check(\"mailbox\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn jaccard_check_len_4() {
        let sample = [
            (true,  "mail"),
            (true,  "bail"),
            (false, "bait"),
            (false, "balt"),
            (false, "bolt"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mail");
            let qtext  = text(query);
            let result = jaccard_check(&rtext.words[0], &qtext.words[0], &rtext.chars, &qtext.chars);
            assert_eq!(result, expect, "Failed jaccard_check(\"mail\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn jaccard_check_len_7() {
        let sample = [
            (true,  "mailbox"),
            (true,  "mailbot"),
            (false, "railbot"),
            (false, "raidbot"),
            (false, "roidbot"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mailbox");
            let qtext  = text(query);
            let result = jaccard_check(&rtext.words[0], &qtext.words[0], &rtext.chars, &qtext.chars);
            assert_eq!(result, expect, "Failed jaccard_check(\"mailbox\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn jaccard_check_len_7_reduction() {
        let sample = [
            (true,  "mailbox"),
            (true,  "mailbxx"),
            (true,  "mailxxx"),
            (false, "maixxxx"),
            (false, "maxxxxx"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mailbox");
            let qtext  = text(query);
            let result = jaccard_check(&rtext.words[0], &qtext.words[0], &rtext.chars, &qtext.chars);
            assert_eq!(result, expect, "Failed jaccard_check(\"mailbox\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn jaccard_check_unfinished() {
        let sample = [
            (true,  "m"),
            (true,  "ma"),
            (true,  "mai"),
            (true,  "mail"),
            (true,  "mailb"),
            (true,  "mailbo"),
            (false, "mailbox"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mail");
            let qtext  = text(query).fin(false);
            let result = jaccard_check(&rtext.words[0], &qtext.words[0], &rtext.chars, &qtext.chars);
            assert_eq!(result, expect, "Failed jaccard_check(\"mail\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn length_check_unfinished() {
        let sample = [
            (true,  "m"),
            (true,  "ma"),
            (true,  "mai"),
            (true,  "mail"),
            (true,  "mailb"),
            (false, "mailbo"),
            (false, "mailbox"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mail");
            let qtext  = text(query).fin(false);
            let result = length_check(&rtext.words[0], &qtext.words[0]);
            assert_eq!(result, expect, "Failed length_check(\"mail\", \"{}\") == {}", query, expect);
        }
    }

    // Match word
    // ----------------------------------------------------------------

    #[test]
    fn match_word_empty_both() {
        let qchars = to_vec("");
        let rchars = to_vec("");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_eq!(word_match(&rword, &qword, &rchars[..], &qchars[..]), None);
    }


    #[test]
    fn match_word_empty_record() {
        let qchars = to_vec("mailbox");
        let rchars = to_vec("");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_eq!(word_match(&rword, &qword, &rchars[..], &qchars[..]), None);
    }


    #[test]
    fn match_word_empty_query() {
        let qchars = to_vec("");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_eq!(word_match(&rword, &qword, &rchars[..], &qchars[..]), None);
    }


    // Match word: prefixed by same length word
    // ----------------------------------------------------------------

    #[test]
    fn match_word_full_strict() {
        let qchars = to_vec("mailbox");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_fuzzy_insertion() {
        let qchars = to_vec("mailybox");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_fuzzy_deletion() {
        let qchars = to_vec("mailox");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_fuzzy_transposition() {
        let qchars = to_vec("maiblox");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_query_too_long() {
        let qchars1 = to_vec("mailboxes");
        let qchars2 = to_vec("mailboxes");
        let rchars  = to_vec("mail");
        let qword1  = Word::new(qchars1.len()).fin(true);
        let qword2  = Word::new(qchars2.len()).fin(false);
        let rword   = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword1, &rchars[..], &qchars1[..]));
        assert_debug_snapshot!(word_match(&rword, &qword2, &rchars[..], &qchars2[..]));
    }

    #[test]
    fn match_word_full_stem() {
        let     rchars  = to_vec("universe");
        let     qchars1 = to_vec("university");
        let     qchars2 = to_vec("university");
        let mut rword   = Word::new(rchars.len());
        let mut qword1  = Word::new(qchars1.len());
        let     qword2  = Word::new(qchars2.len());

        let lang = lang_english();
        qword1.stem(&qchars1[..], &lang);
        rword .stem(&rchars[..],  &lang);

        assert_debug_snapshot!(word_match(&rword, &qword1, &rchars[..], &qchars1[..]));
        assert_debug_snapshot!(word_match(&rword, &qword2, &rchars[..], &qchars2[..]));
    }


    // Match word: prefixed by lesser length word
    // ----------------------------------------------------------------

    #[test]
    fn match_word_partial_strict() {
        let qchars = to_vec("mailb");
        let rchars = to_vec("mailbox");
        let qword = Word::new(qchars.len()).fin(false);
        let rword = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_partial_fuzzy_insertion() {
        let qchars = to_vec("maiylb");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_partial_fuzzy_deletion() {
        let qchars = to_vec("maib");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_partial_fuzzy_transposition() {
        let qchars = to_vec("malib");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }
}
