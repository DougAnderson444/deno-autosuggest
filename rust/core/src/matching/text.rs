use fnv::{FnvHashSet as HashSet};
use crate::tokenization::Text;
use super::WordMatch;
use super::word::word_match;


pub fn text_match(rtext: &Text<&[char]>, qtext: &Text<&[char]>) -> Vec<WordMatch> {
    let rchars   = &rtext.chars;
    let qchars   = &qtext.chars;
    let capacity = min!(rtext.words.len(), qtext.words.len());
    let mut rtaken: HashSet<usize>  = HashSet::with_capacity_and_hasher(capacity, Default::default());
    let mut qtaken: HashSet<usize>  = HashSet::with_capacity_and_hasher(capacity, Default::default());
    let mut matches: Vec<WordMatch> = Vec::with_capacity(capacity);

    for qword in qtext.words.iter() {
        if qtaken.contains(&qword.ix) { continue; }

        let mut candidate: Option<(WordMatch, Option<WordMatch>)> = None;

        for rword in rtext.words.iter() {
            if rtaken.contains(&rword.ix) { continue; }

            let new_candidate = None
                .or_else(|| {
                    let rnext = rtext.words.get(rword.ix + 1)?;
                    if qword.len() <= rword.len() + rword.dist(rnext) { return None; }
                    if rtaken.contains(&(rword.ix + 1)) { return None; }
                    let m        = word_match(&rword.join(rnext), qword, rchars, qchars)?;
                    let (m1, m2) = m.split_record(rword, rnext);
                    Some((m1, Some(m2)))
                })
                .or_else(|| {
                    let qnext = qtext.words.get(qword.ix + 1)?;
                    if rword.len() <= qword.len() + qword.dist(qnext) { return None; }
                    if qtaken.contains(&(qword.ix + 1)) { return None; }
                    let m        = word_match(rword, &qword.join(qnext), rchars, qchars)?;
                    let (m1, m2) = m.split_query(qword, qnext);
                    Some((m1, Some(m2)))
                })
                .or_else(|| {
                    let m = word_match(rword, qword, rchars, qchars)?;
                    Some((m, None))
                });

            if let Some(new_candidate) = new_candidate {
                if !new_candidate.0.record.function || new_candidate.1.is_some() {
                    candidate = Some(new_candidate);
                    break;
                }
                if !candidate.is_some() {
                    candidate = Some(new_candidate);
                    continue;
                }
            }
        }

        if let Some((m1, m2)) = candidate {
            qtaken.insert(m1.query.ix);
            rtaken.insert(m1.record.ix);
            matches.push(m1);
            if let Some(m2) = m2 {
                qtaken.insert(m2.query.ix);
                rtaken.insert(m2.record.ix);
                matches.push(m2);
            }
        }
    }

    matches
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::tokenization::Text;
    use crate::lang::{Chars, lang_english};
    use super::{text_match};


    fn text(s: &str) -> Text<Vec<char>> {
        Text::from_str(s).split(&[Chars::Punctuation, Chars::Whitespaces])
    }


    #[test]
    fn match_text_empty_both() {
        let qtext = text("").fin(false);
        let rtext = text("");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_empty_one() {
        let qtext = text("mailbox").fin(false);
        let rtext = text("");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
        assert_debug_snapshot!(text_match(&qtext.to_ref(), &rtext.to_ref()));
    }


    #[test]
    fn match_text_singleton_equality() {
        let qtext = text("mailbox").fin(false);
        let rtext = text("mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_singleton_typos() {
        let qtext = text("maiblox").fin(false);
        let rtext = text("mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_pair_first() {
        let qtext = text("yelow").fin(false);
        let rtext = text("yellow mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_pair_second() {
        let qtext = text("maiblox").fin(false);
        let rtext = text("yellow mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_pair_unfinished() {
        let qtext = text("maiblox yel").fin(false);
        let rtext = text("yellow mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_intersection() {
        let qtext = text("big malibox yelo").fin(false);
        let rtext = text("small yellow metal mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_best_rword() {
        let mut qtext = text("the").fin(false);
        let mut rtext = text("the theme");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));

        let lang = Some(lang_english());
        qtext = qtext.mark_pos(&lang);
        rtext = rtext.mark_pos(&lang);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_regression_best_match() {
        let qtext = text("sneak").fin(false);
        let rtext = text("sneaky");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_query() {
        let qtext = text("wi fi router").fin(false);
        let rtext = text("wifi router");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_query_unfihished() {
        let qtext = text("micro bio").fin(false);
        let rtext = text("microbiology");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_query_typos() {
        let qtext = text("mcro byology").fin(false);
        let rtext = text("microbiology");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_record() {
        let qtext = text("wifi router").fin(false);
        let rtext = text("wi fi router");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_record_typos() {
        let qtext = text("mcrobiology").fin(false);
        let rtext = text("micro biology");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_record_unfinished() {
        let qtext = text("microbio").fin(false);
        let rtext = text("micro biology");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_regression_1() {
        let qtext = text("especiall").fin(false);
        let rtext = text("special, year");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }
}
