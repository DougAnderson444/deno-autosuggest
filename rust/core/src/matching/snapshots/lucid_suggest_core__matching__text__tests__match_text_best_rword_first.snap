---
source: src/matching/text.rs
expression: "text_match(&rtext.to_ref(), &qtext.to_ref())"
---
(
    [
        WordMatch {
            ix: 0,
            len: 6,
            slice: (
                0,
                3,
            ),
            typos: 0.0,
            func: false,
            fin: false,
        },
    ],
    [
        WordMatch {
            ix: 0,
            len: 3,
            slice: (
                0,
                3,
            ),
            typos: 0.0,
            func: false,
            fin: false,
        },
    ],
)
